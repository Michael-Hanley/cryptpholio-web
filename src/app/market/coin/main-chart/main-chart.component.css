.spinner {
  margin-top: 125px;
  width: 350px;
  height: 350px;
}
.chartToggle {
  margin-left: 15px;
  margin-bottom: 1px;
}