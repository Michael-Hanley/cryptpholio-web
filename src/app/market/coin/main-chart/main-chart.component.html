<div *ngIf="ready">
  <div class="chartToggle">
    <mat-button-toggle-group #group="matButtonToggleGroup">
      <mat-button-toggle *ngFor="let timeline of timelineOptions" value="{{timeline}}" [checked]="timeline === currentTimeline" (click)="currentTimeline = timeline; getCoinHistory(); ready = false">
        {{timeline}}
      </mat-button-toggle>
    </mat-button-toggle-group>
  </div>
  <div [chart]="chart"></div>
</div>
<div fxLayout fxLayoutAlign="center center" >
  <mat-spinner class="spinner" *ngIf="!ready">
  </mat-spinner>
</div>
