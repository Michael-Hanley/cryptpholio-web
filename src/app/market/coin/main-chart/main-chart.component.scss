.spinner {
  margin-top: 125px;
  width: 350px;
  height: 350px;
}
.chartToggle {
  margin: 15px;
}

.highcharts-background {
  background-color: transparent;
}