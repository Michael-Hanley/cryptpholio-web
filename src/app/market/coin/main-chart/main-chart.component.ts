import { Component, OnInit, Input, OnDestroy } from '@angular/core';

import { Chart } from 'angular-highcharts';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

import { CoinService } from '../../../services/coin.service';
import { ThemeService } from '../../../services/theme.service';

@Component({
  selector: 'app-main-chart',
  templateUrl: './main-chart.component.html',
  styleUrls: ['./main-chart.component.scss']
})
export class MainChartComponent implements OnInit, OnDestroy {
  @Input() symbol;
  @Input() supply;
  @Input() name;

  history;
  chartHistory;
  chart;
  ready = false;
  timelineOptions = ['1h', '24h', '7d', '1m', '6m', '1y', '3y'];
  currentTimeline = '3y';
  theme;

  private ngUnsubscribe: Subject<any> = new Subject();

  constructor(private coinService: CoinService, private themeService: ThemeService) {}

  createGraph() {
    this.chart = new Chart({
      chart: {
        type: 'line',
        backgroundColor: this.theme.darkColor
      },
      title: {
        text: this.name,
        style: {
          color: this.theme.lightColor
        }
      },
      legend: {
        itemStyle: {
          color: this.theme.lightColor
        }
      },
      xAxis: {
        type: 'datetime',
        title: {
          text: 'Date',
          style: {
            color: this.theme.lightColor
          }
        },
        labels: {
          style: {
            //
            color: this.theme.lightColor
          },
          format: '{value:%e-%b-%Y}'
        }
      },
      yAxis: {
        title: {
          text: 'Market Cap',
          style: {
            color: this.theme.lightColor
          }
        },
        labels: {
          style: {
            color: this.theme.lightColor
          }
        }
      },
      credits: {
        enabled: false
      },
      plotOptions: {
        series: {
          turboThreshold: 5000
        }
      },
      series: [{
        type: 'line',
        name: this.symbol,
        data: this.chartHistory,
        dataLabels: {
          style: {
            color: this.theme.lightColor
          }
        }
      }],
    });
  }

  getCoinHistory() {
    this.coinService.getHistory(this.symbol, this.currentTimeline)
    .pipe(takeUntil(this.ngUnsubscribe))
    .subscribe(hist => {
      this.chartHistory = this.createGraphData(hist);
      this.createGraph();
      this.ready = true;
    });
  }

  createGraphData(history) {
    const market_cap_usd_arr = [];

    history.Data.forEach(time => {
      const date = new Date(0);

      market_cap_usd_arr.push({
        y: this.coinService.calculateMarketCap(time.high, this.supply),
        x: date.setUTCSeconds(time.time)
      });
    });

    return market_cap_usd_arr;
  }

  ngOnInit() {
    this.getCoinHistory();
    this.themeService._themeObject$
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe(theme => {
        this.theme = theme;
        this.createGraph();
      });
  }

  ngOnDestroy() {
      this.chart = null;
      this.ngUnsubscribe.next();
      this.ngUnsubscribe.complete();
  }
}

