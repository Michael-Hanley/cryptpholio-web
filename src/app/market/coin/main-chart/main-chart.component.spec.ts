import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule } from '@angular/common/http/testing';

import { MainChartComponent } from './main-chart.component';

import { MatButtonToggleModule } from '@angular/material/button-toggle';
import { ChartModule } from 'angular-highcharts';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';

describe('MainChartComponent', () => {
  let component: MainChartComponent;
  let fixture: ComponentFixture<MainChartComponent>;
  const history = {
    Data: [
      {high: 100.00, time: 1578960000},
      {high: 200.00, time: 1578970000}
    ]
  };
  
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ MainChartComponent ],
      imports: [
        MatButtonToggleModule,
        ChartModule,
        MatProgressSpinnerModule,
        HttpClientTestingModule
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MainChartComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should create graph data', () => {
    component.supply = 100000;
    expect(component.createGraphData(history)).toMatchObject([{'x': 1578960000000, 'y': 10000000}, {'x': 1578970000000, 'y': 20000000}]);
  });
  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
