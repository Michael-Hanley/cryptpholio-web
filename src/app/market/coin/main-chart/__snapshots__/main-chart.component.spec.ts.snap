// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainChartComponent should match snapshot 1`] = `
<app-main-chart
  chart={[Function Chart]}
  coinService={[Function CoinService]}
  currentTimeline={[Function String]}
  ngUnsubscribe={[Function Subject]}
  ready="false"
  theme={[Function Object]}
  themeService={[Function ThemeService]}
  timelineOptions={[Function Array]}
>
  <div>
    <div
      class="chartToggle"
    >
      <mat-button-toggle-group
        aria-disabled="false"
        class="mat-button-toggle-group mat-button-toggle-group-appearance-standard"
        role="group"
      >
        
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-17"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-17-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               1h 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-18"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-18-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               24h 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-19"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-19-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               7d 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-20"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-20-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               1m 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-21"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-21-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               6m 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-appearance-standard"
          id="mat-button-toggle-22"
          tabindex="-1"
        >
          <button
            aria-pressed="false"
            class="mat-button-toggle-button"
            id="mat-button-toggle-22-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               1y 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
        <mat-button-toggle
          class="mat-button-toggle mat-button-toggle-checked mat-button-toggle-appearance-standard"
          id="mat-button-toggle-23"
          tabindex="-1"
        >
          <button
            aria-pressed="true"
            class="mat-button-toggle-button"
            id="mat-button-toggle-23-button"
            name="mat-button-toggle-group-16"
            tabindex="0"
            type="button"
          >
            <div
              class="mat-button-toggle-label-content"
            >
               3y 
            </div>
          </button>
          <div
            class="mat-button-toggle-focus-overlay"
          />
          <div
            class="mat-button-toggle-ripple mat-ripple"
            matripple=""
          />
        </mat-button-toggle>
      </mat-button-toggle-group>
    </div>
    
  </div><div
    fxlayout=""
    fxlayoutalign="center center"
  >
    
    <mat-spinner
      class="spinner mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation"
      mode="indeterminate"
      role="progressbar"
      style="width: 100px; height: 100px;"
    >
      <svg
        focusable="false"
        preserveAspectRatio="xMidYMid meet"
        style="width: 100px; height: 100px;"
        viewBox="0 0 100 100"
      >
        
        <circle
          cx="50%"
          cy="50%"
          r="45"
          style="animation-name: mat-progress-spinner-stroke-rotate-100; stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
        />
        
      </svg>
    </mat-spinner>
  </div>
</app-main-chart>
`;
